qiita_db/processing_job.py

```diff
@@ -6,21 +6,291 @@
 # -----------------------------------------------------------------------------
 
-from uuid import UUID
+import networkx as nx
+import qiita_db as qdb
+from collections import defaultdict, Iterable
 from datetime import datetime
-from subprocess import Popen, PIPE
-from multiprocessing import Process
-from os.path import join
+from future.utils import viewitems, viewvalues
 from itertools import chain
-from collections import defaultdict, Iterable
 from json import dumps, loads
+from multiprocessing import Process, Queue, Event
+from qiita_core.qiita_settings import qiita_config
+from qiita_db.util import create_nested_path
+from re import search, findall
+from subprocess import Popen, PIPE
 from time import sleep
+from uuid import UUID
+from os.path import join
 
-from future.utils import viewitems, viewvalues
-import networkx as nx
 
-from qiita_core.qiita_settings import qiita_config
-import qiita_db as qdb
+class Watcher(Process):
+    # TODO: Qiita will need a proper mapping of these states to Qiita states
+    # Currently, these strings are being inserted directly into Qiita's status
+    # table. Qiita will be unfamiliar with many of these. We will need at least
+    # one additional job type for 'Held': A job waiting for another to complete
+    # before it can run.
+    #
+    # Note that the main Qiita script instantiates an object of this class in
+    # a separate thread, so it can periodically update the database w/metadata
+    # from Watcher's queue. Qiita's script also calls qdb.complete() so there
+    # are no circular references. TODO: replace w/a REST call.
+
+    # valid Qiita states:
+    #             The current status of the job, one of {'queued', 'running',
+    #             'success', 'error', 'in_construction', 'waiting'}
+
+    # TODO: what to map in_construction to?
+    job_state_map = {'C': 'completed', 'E': 'exiting', 'H': 'held',
+                     'Q': 'queued', 'R': 'running', 'T': 'moving',
+                     'W': 'waiting', 'S': 'suspended'}
+
+    torque_to_qiita_state_map = {'completed': 'completed',
+                                 'exiting': 'running',
+                                 'held': 'queued',
+                                 'queued': 'queued',
+                                 'running': 'running',
+                                 'moving': 'waiting',
+                                 'waiting': 'waiting',
+                                 'suspended': 'waiting',
+                                 'DROPPED': 'error'}
+
+    def __init__(self):
+        super(Watcher, self).__init__()
+
+        # set self.owner to qiita, or whomever owns processes we need to watch.
+        self.owner = 'qiita@qiita.ucsd.edu'
+
+        # Torque is set to drop jobs from its queue 60 seconds after
+        # completion, by default. Setting a polling value less than
+        # that allows for multiple chances to catch the exit status
+        # before it disappears.
+        self.polling_value = 15
```

Comment:
Jeff mentioned that he's open to adjusting this value to suit our needs, and 60 seconds is just the default. Since it's easy to adjust the value on Torque's end, it really should be easy to change on Qiita's end as well. I think putting it in the db is a solid idea.